pkg/tools/prefix_pool/prefixpool.go

```diff
@@ -0,0 +1,642 @@
+	}
+
+	return &net.IPNet{
+		IP:   setNetIndexInIP(ipnet.IP, subnet_index, newPrefixLen),
+		Mask: net.CIDRMask(newPrefixLen, addrLen),
+	}, nil
+}
+
+func AddressCount(prefixes ...string) uint64 {
+	var c uint64 = 0
+	for _, pr := range prefixes {
+		c += addressCount(pr)
+	}
+	return c
+}
+
+func addressCount(pr string) uint64 {
+	_, network, _ := net.ParseCIDR(pr)
+	prefixLen, bits := network.Mask.Size()
+	return 1 << (uint64(bits) - uint64(prefixLen))
+}
+
+func setNetIndexInIP(ip net.IP, num int, prefixLen int) net.IP {
+	ipInt, totalBits := fromIP(ip)
+	bigNum := big.NewInt(int64(num))
+	bigNum.Lsh(bigNum, uint(totalBits-prefixLen))
+	ipInt.Or(ipInt, bigNum)
+	return toIP(ipInt, totalBits)
+}
+
+func clearNetIndexInIP(ip net.IP, prefixLen int) net.IP {
+	ipInt, totalBits := fromIP(ip)
+	ipInt.SetBit(ipInt, totalBits-prefixLen, 0)
+	return toIP(ipInt, totalBits)
+}
+
+func toIP(ipInt *big.Int, bits int) net.IP {
+	ipBytes := ipInt.Bytes()
+	ret := make([]byte, bits/8)
+	// Pack our IP bytes into the end of the return array,
+	// since big.Int.Bytes() removes front zero padding.
+	for i := 1; i <= len(ipBytes); i++ {
+		ret[len(ret)-i] = ipBytes[len(ipBytes)-i]
+	}
+	return net.IP(ret)
+}
+
+func fromIP(ip net.IP) (*big.Int, int) {
+	val := &big.Int{}
+	val.SetBytes([]byte(ip))
+	i := len(ip)
+	if i == net.IPv4len {
+		return val, 32
+	} // else if i == net.IPv6len
+	return val, 128
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func MaxCommonPrefixSubnet(s1, s2 *net.IPNet) *net.IPNet {
+	rawIp1, n1 := fromIP(s1.IP)
+	rawIp2, _ := fromIP(s2.IP)
+
+	xored := &big.Int{}
+	xored.Xor(rawIp1, rawIp2)
+	maskSize := leadingZeros(xored, n1)
+
+	m1, bits := s1.Mask.Size()
+	m2, _ := s2.Mask.Size()
+
+	mask := net.CIDRMask(min(min(m1, m2), maskSize), bits)
+	return &net.IPNet{
+		IP:   s1.IP.Mask(mask),
+		Mask: mask,
+	}
+}
+
+func IpToNet(ipAddr net.IP) *net.IPNet {
+	mask := net.CIDRMask(len(ipAddr)*8, len(ipAddr)*8)
+	return &net.IPNet{IP: ipAddr, Mask: mask}
+}
+
+func leadingZeros(n *big.Int, size int) int {
+	i := size - 1
+	for ; n.Bit(i) == 0 && i > 0; i-- {
+	}
+	return size - 1 - i
+}
+
+/**
+AddressRange returns the first and last addresses in the given CIDR range.
+*/
+func AddressRange(network *net.IPNet) (net.IP, net.IP) {
+	firstIP := network.IP
+	// the last IP is the network address OR NOT the mask address
+	prefixLen, bits := network.Mask.Size()
+	if prefixLen == bits {
+		lastIP := make([]byte, len(firstIP))
+		copy(lastIP, firstIP)
+		return firstIP, lastIP
+	}
+
+	firstIPInt, bits := fromIP(firstIP)
+	hostLen := uint(bits) - uint(prefixLen)
+	lastIPInt := big.NewInt(1)
+	lastIPInt.Lsh(lastIPInt, hostLen)
+	lastIPInt.Sub(lastIPInt, big.NewInt(1))
+	lastIPInt.Or(lastIPInt, firstIPInt)
+
+	return firstIP, toIP(lastIPInt, bits)
+}
+
+func IncrementIP(sourceIp net.IP, ipNet *net.IPNet) (net.IP, error) {
```

Comment:
Can we make this thing private?

Reply:
This function is used in another package `test/kubetest`

Reply:
So let's make this function private

Reply:
Made the function private, it's used also in prefixpool, so let it be :)